smt2parser/src/parser.rs

```diff
@@ -355,6 +359,18 @@ pomelo! {
     command ::= LeftParen SetLogic bound_symbol(x) RightParen { extra.0.visit_set_logic(x)? }
     //   ( set-option ⟨attribute⟩ )
     command ::= LeftParen SetOption keyword(k) attribute_value(v) RightParen { extra.0.visit_set_option(k, v)? }
+
+
+    //   ( theory ⟨symbol⟩ ⟨theory_attributes⟩+ )
+    theory ::= LeftParen Theory fresh_symbol(s) RightParen { extra.0.visit_theory(s)? }
+
+    mode ::= command(c) { ParseResult::Command(c) }
+    mode ::= theory(td) { ParseResult::Theory(td) }
+}
+
+pub enum ParseResult<C, T> {
```

Comment:
This seems reasonable. (Need inline doc comment for public API though :))
In full generality, we could pass an artificial input token to signal what we expect but this is not even needed here since there is no ambiguity.